dev/com.ibm.ws.security.authorization.jacc.ejb/src/com/ibm/ws/security/authorization/jacc/ejb/impl/EJBSecurityValidatorImpl.java

```diff
@@ -84,55 +98,70 @@ private boolean checkMethodConstraints(final String contextId,
-                                MessageContext mc = null;
-                                try {
-                                    InitialContext ic = new InitialContext();
-                                    mc = getMessageContext(ic);
-                                } catch (NamingException e) {
-                                    if (tc.isDebugEnabled())
-                                        Tr.debug(tc, "NamingException is caught. Ignoring..", e);
-                                }
-                                if (mc != null) {
-                                    if (tc.isDebugEnabled())
-                                        Tr.debug(tc, "javax.xml.soap.SOAPMessage is set: ", mc);
-                                    handlerObjects.put(jaccHandlerKeyArray[3], mc);
-                                }
-                                ProtectionDomain pd = null;
-
-                                if (subject != null && subject.getPrincipals().size() > 0) {
-                                    Principal[] principalArray = subject.getPrincipals().toArray(new Principal[subject.getPrincipals().size()]);
-                                    pd = new ProtectionDomain(nullCs, null, null, principalArray);
-                                } else {
-                                    pd = nullPd;
-                                }
-
-                                if (tc.isDebugEnabled())
-                                    Tr.debug(tc, "Setting JACC handler data");
-                                PolicyContext.setHandlerData(handlerObjects);
-                                if (tc.isDebugEnabled())
-                                    Tr.debug(tc, "Calling JACC implies. PD : " + pd);
-                                return Policy.getPolicy().implies(pd, permission);
-                            }
-                        });
+                                               new PrivilegedExceptionAction<Boolean>() {
+                                                   @Override
+                                                   public Boolean run() throws javax.security.jacc.PolicyContextException {
+                                                       PolicyContext.setContextID(contextId);
+
+                                                       if (tc.isDebugEnabled())
+                                                           Tr.debug(tc, "Registering JACC context handlers");
+
+                                                       for (String key : jaccHandlerKeyArrayEe8) {
+                                                           PolicyContext.registerHandler(key, pch, true);
+                                                       }
+                                                       for (String key : jaccHandlerKeyArrayEe9) {
+                                                           PolicyContext.registerHandler(key, pch, true);
```

Comment:
Is PolicyContextHandlerImpl the only possible handler?  By setting both you can get an IllegalArgumentException if the handler doesn't support the key.  If it is a JACC 1.5 handler it won't know how to process the EE9 arguments, so you will end up breaking zero migration.  Also since you know that argument 0 and 1 are the same it seems like a waste to call it for each one of them even if you stuck with this approach.  I think you need to do what you were doing before and just do the first arguments for the spec level.  The jacc provider will have to implement the jarkarta handler and update to do the new parameters anyway.  Maybe have a common array as well if you are going to keep with your current approach.

Reply:
It is the only PolicyContextHandler Liberty provides.

I had talked to @kwsutter and we preferred to support both sets of properties for ease of use in migration. When I start working on the FATs I will keep an eye on any potential issues that might arise from supporting both, but my suspicion is that the ee8- providers only look for the javax properties so the new properties won't be an issue.

As for the array entries that I am essentially overwriting, the reason I put them in there is because when the spec is updated and we come back and add the updated keys, all we will need to do is update the array and the code to set the key will already be there.